File: tools/zexcmds.py
"""Every Ex command name, typed at `:`, and what the editor said.

Usage: python3 tools/zexcmds.py <binary> <source> <outfile>

`tools/exsweep.py` recorded three things zero does not have: an exit status worth
reading (the editor is quit by the harness, not by the command), the files left
in the cwd, and stderr.  What is left is what the editor SAYS, and the message
line is recordable because `tools/zscreen.py` snapshots every redraw -- with only
the final screen, every row here read `~`, because the keys that quit wipe the
message (measured).

So a row is the screen after the command: the text at the top, and every distinct
message line from the command onwards.  That is a message-level record where the
old one was an exit status -- retiring `:write` moves `E32: No file name` to
`E492: Not an editor command`, which the file sweep could not have seen, both
being exit 1.

Two things it had to be taught, both measured (ZERO-PLAN.md 2i):

  * `:highlight` pages, and ESC does not stop a `--More--` listing -- the session
    then ran to the timeout.  `q` ends it, and in Normal mode `q` followed by ESC
    is an aborted recording that changes nothing.  (`.` also works and repeats
    the last change, which polluted the row.)
  * `:stop` and `:suspend` hand the process group to the shell; they are skipped
    here as they are in the file sweep, and every run gets a session of its own
    anyway (tools/zstream.py).
"""
import concurrent.futures
import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import create_cmdidxs as cmdidxs
import zrec
import zstream

SEED = b'alpha\rbeta\rgamma'
SKIP = {'stop', 'suspend'}


def one(args):
    name, binary = args
    if name in SKIP:
        return '=== %s\n%s' % (name, zrec.section('skipped', 'hands over the process group'))
    keys = [b'i' + SEED + b'\x1b', b':set nopaste\r',
            b':' + name.encode() + b'\r',
            b'q', b'\x1b', b':q!\r']
    try:
        scr, out, err, rc = zstream.session(binary, keys, args=['+set paste'])
    except zstream.Blocked:
        return '=== %s\n%s' % (name, zrec.section('blocked', 'never returned'))
    text = ''
    msgs = []
    for i, (dump, y, x, bells) in enumerate(scr.snaps):
        lines = dump.split('\n')
        # The seed and the `:set nopaste` make the first snapshots; from the
        # command's own onwards, the top of the screen and every distinct last row.
        if i >= 2:
            if not text:
                text = ' | '.join(l for l in lines[:6] if l.strip() and l.strip() != '~')
            m = lines[-1].rstrip()
            if m and m not in msgs:
                msgs.append(m)
    row = '=== %s\n' % name
    row += zrec.section('exit %s' % rc)
    row += zrec.section('bells %d' % scr.bells)
    row += zrec.section('stderr', err.decode('utf-8', 'replace').rstrip('\n'))
    row += zrec.section('text', text)
    row += zrec.section('msgs', '\n'.join(msgs))
    return zrec.scrub(row)


def main():
    binary, table, out = (os.path.abspath(sys.argv[1]), os.path.abspath(sys.argv[2]),
                          os.path.abspath(sys.argv[3]))
    names = cmdidxs.names(table)
    t = time.time()
    with concurrent.futures.ThreadPoolExecutor(
            max_workers=min(len(names), (os.cpu_count() or 4) * 2)) as ex:
        rows = list(ex.map(one, [(n, binary) for n in names]))
    open(out, 'w').write(''.join(rows))
    print('%d commands -> %s in %.1fs' % (len(rows), out, time.time() - t))


if __name__ == '__main__':
    main()

File: tools/go/internal/harness/zexcmds.go
package harness

import (
	"fmt"
	"io"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"
)

// zexcmdsSeed is what every command is given to work on.
var zexcmdsSeed = "alpha\rbeta\rgamma"

// zexcmdsSkip hand the process group to the shell.  They are skipped here as
// they are in the file sweep, although every run gets a session of its own
// anyway.
var zexcmdsSkip = map[string]bool{"stop": true, "suspend": true}

// ZExCmds types every Ex command at ':' and records the message it printed.
//
// Two details in the key sequence are load-bearing.  `:highlight` pages and
// ESC does NOT stop a --More-- listing, so a session without an answer for it
// ran to the timeout; `q` ends the listing, and in Normal mode `q` followed by
// ESC is an aborted recording that changes nothing.  (`.` also ends it, and
// repeats the last change, which polluted the row.)
//
// The editor starts with +set paste and the seed is typed under it, then
// 'nopaste' is set back: the corpus types its own text rather than opening a
// file, because from zero phase 8 there is no way to name one.
func ZExCmds(bin, table, out string, w io.Writer) error {
	names, err := CommandNames(table)
	if err != nil {
		return err
	}
	start := time.Now()
	rows := make([]string, len(names))
	n := runtime.NumCPU() * 2
	if n > len(names) {
		n = len(names)
	}
	sem := make(chan struct{}, n)
	var wg sync.WaitGroup
	for i, name := range names {
		wg.Add(1)
		go func(i int, name string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			rows[i] = zexcmdsOne(bin, name)
		}(i, name)
	}
	wg.Wait()
	if err := os.WriteFile(out, []byte(strings.Join(rows, "")), 0o644); err != nil {
		return err
	}
	fmt.Fprintf(w, "%d commands -> %s in %.1fs\n", len(rows), out, time.Since(start).Seconds())
	return nil
}

func zexcmdsOne(bin, name string) string {
	if zexcmdsSkip[name] {
		body := "hands over the process group"
		return "=== " + name + "\n" + Section("skipped", &body)
	}
	keys := [][]byte{
		[]byte("i" + zexcmdsSeed + "\x1b"),
		[]byte(":set nopaste\r"),
		[]byte(":" + name + "\r"),
		[]byte("q"), []byte("\x1b"), []byte(":q!\r"),
	}
	scr, _, stderr, rc, err := ZSession(bin, keys, "xterm", []string{"+set paste"},
		24, 80, 20*time.Second)
	if err != nil {
		body := "never returned"
		return "=== " + name + "\n" + Section("blocked", &body)
	}

	text := ""
	var msgs []string
	for i, snap := range scr.Snaps {
		lines := strings.Split(snap.Text, "\n")
		// The seed and the `:set nopaste` make the first two snapshots; from
		// the command's own onwards, the top of the screen and every distinct
		// last row.
		if i < 2 {
			continue
		}
		if text == "" {
			var keep []string
			hi := 6
			if len(lines) < hi {
				hi = len(lines)
			}
			for _, l := range lines[:hi] {
				t := strings.TrimSpace(l)
				if t != "" && t != "~" {
					keep = append(keep, l)
				}
			}
			text = strings.Join(keep, " | ")
		}
		m := strings.TrimRight(lines[len(lines)-1], " \t\n\v\f\r")
		if m != "" && !containsStr(msgs, m) {
			msgs = append(msgs, m)
		}
	}

	row := "=== " + name + "\n"
	row += Section(fmt.Sprintf("exit %d", rc), nil)
	row += Section(fmt.Sprintf("bells %d", scr.Bells), nil)
	errBody := strings.TrimRight(decodeReplace(stderr), "\n")
	row += Section("stderr", &errBody)
	row += Section("text", &text)
	msgBody := strings.Join(msgs, "\n")
	row += Section("msgs", &msgBody)
	return Scrub(row)
}

func containsStr(xs []string, s string) bool {
	for _, x := range xs {
		if x == s {
			return true
		}
	}
	return false
}
